Add concentration_tracker component and tests.

loroberge opened the pull request:
### Description

This PR addresses Issue #1662 . It adds the _ConcentrationTrackerForDiffusion_ component, which calculates the concentration of user-defined sediment properties for hillslope sediments. 

The component must be coupled with either _DepthDependentDiffuser_ or _DepthDependentTaylorDiffuser_, as it attaches concentrations to the _"soil__flux"_ field those components generate and then redistributes the concentrations according to a flux divergence calculation in the exact manner of those two components. I explain this in the docstring but don't have anything in the code that checks/warns about using a different flux component. If the _"soil__flux"_ field was calculated in another manner (e.g., allowing for diagonal fluxes), the concentration tracker will run just fine but redistribute concentrations in a manner inconsistent with the true movement of the sediment.

At the moment, the component can only track one concentration field. I intend to add the option to track multiple fields in the coming months.

I have added some tests to check the different user input options. I don't know what else I should cover. Please let me know if I'm missing anything.

### Checklist - did you ...

<!-- If any of the following items aren't relevant for your contribution
     please still tick them so we know you've gone through the checklist.

    Tests are required for bugfixes and new features. Documentation changes
    are necessary for formatting and most enhancement changes.

    Ensure proper code formatting with black. You can do this by running
    black from landlab's top-level folder.

    Ensure landlab is lint-free by running flake8 at landlab's top-level
    folder.

    If you like, you can automate these tasks by installing pre-commit hooks.
    This is done by running `pre-commit install` at landlab's top-level
    folder.
-->

- [x] Add a [news fragment](https://landlab.readthedocs.io/en/master/development/contribution/index.html#news-entries) file entry if necessary?
- [x] Add / update tests if necessary?
- [x] Add new / update outdated documentation?
- [x] All tests have passed?
- [x] Formatted code with black?
- [x] Removed lint reported by flake8?
- [x] Sucessful documentation built? (if documentation added or modified)

<!-- Thanks for your time and effort. If you have any feedback in regards
     to your experience contributing here, please let us know!

     Helpful links:

      Developer guide: https://landlab.readthedocs.io/en/master/development
      Ask a question or submit an issue: https://github.com/landlab/landlab/issues
      Landlab Slack channel: https://landlab.slack.com
-->

loroberge commented:
@mcflugen I need some help with a couple issues:

1. Documentation / Build documentation fails with  **WARNING: document isn't included in any toctree**
I see that I should add the component to landlab/docs/source/reference/components/index.rst. Is there a specific category in which I should place the component? 
2. Lint / Check for lint fails with exit code 1. I am unable to run nox on my local build for some reason. I am not sure how to solve this lint check.

mcflugen commented:
> I am unable to run nox on my local build for some reason.

What is the error you are seeing?

> Is there a specific category in which I should place the component?

Since it works with the `DepthDependentDiffuser`, maybe *Hillslope Geomorphology*? What do you think?

loroberge commented:
> > I am unable to run nox on my local build for some reason.
> 
> What is the error you are seeing?

I get the following:
```
ERROR: Could not find a version that satisfies the requirement pre-commit (from versions: none)
ERROR: No matching distribution found for pre-commit
WARNING: pip is configured with locations that require TLS/SSL, however the ssl module in Python is not available.
```

> > Is there a specific category in which I should place the component?
> 
> Since it works with the `DepthDependentDiffuser`, maybe _Hillslope Geomorphology_? What do you think?

Sounds good. That's also what I was thinking.

mcflugen commented:
@loroberge That error message seems to indicate that openssl isn't install within your python distribution. What Python version and OS are you using? Are you able to *pip* install anything at all (e.g. does `pip install pre-commit` work)?

loroberge commented:
> @loroberge That error message seems to indicate that openssl isn't install within your python distribution. What Python version and OS are you using? Are you able to _pip_ install anything at all (e.g. does `pip install pre-commit` work)?

I am using windows with python 3.9.17. 

I have pip installed other things in the past. Trying `pip install pre-commit` returns many lines telling me:
`Requirement already satisfied: ` for various packages

loroberge commented:
@mcflugen  I have solved the SSL problem and nox is working fine.

All checks now pass except for some notebooks unrelated to this PR. 

@mcflugen or @gregtucker does one of you have time this week for a zoom meeting? I would like to know 1. whether my tests cover everything and 2. if they are doing what I intend.

loroberge commented:
@mcflugen I think this PR is good to go. Can you review?

mcflugen commented on landlab/components/concentration_tracker/concentration_tracker_for_diffusion.py:
```diff
@@ -0,0 +1,521 @@
+        dQCdx = self._grid.calc_flux_div_at_node(self._grid.at_link["QC"])
+
+        # Calculate other components of mass balance equation
+        with np.errstate(divide="ignore", invalid="ignore"):
+            C_local = C_old * (self._soil__depth_old / self._soil__depth)
+            C_from_weathering = (
+                self._C_br * (self._soil_prod_rate * dt) / self._soil__depth
+            )
+            Production = (dt * self._P / 2) * (
+                self._soil__depth_old / self._soil__depth + 1
+            )
+            Decay = (dt * self._D / 2) * (self._soil__depth_old / self._soil__depth + 1)
+
+            # Calculate concentration
+            self._concentration[:] = (
+                C_local
+                + C_from_weathering
+                + (dt / self._soil__depth) * (-dQCdx)
+                + Production
+                - Decay
+            )
```
Instead of of changing the `np.errstate` could you, instead, be more explicit about what to do with nodes with zero soil depth? Maybe something along the lines of,
```python
is_soil = self._soil__depth > 0.0

C_local = np.divide(
    C_old * self._soil__depth_old, self._soil__depth, where=is_soil
)
C_from_weathering = np.divide(
    self._C_br * (self._soil_prod_rate * dt), self._soil__depth, where=is_soil
)
Production = (dt * self._P / 2) * (
    np.divide(self._soil__depth_old, self._soil__depth, where=is_soil) + 1
)
Decay = (dt * self._D / 2) * (
    np.divide(self._soil__depth_old, self._soil__depth, where=is_soil) + 1
)

self._concentration[:] = (
    C_local
    + C_from_weathering
    + (dt / self._soil__depth) * (-dQCdx)
    + Production
    - Decay
)

self._concentration[~is_soil] = 0.0
```

mcflugen commented on tests/components/concentration_tracker/test_concentration_tracker_for_diffusion.py:
```diff
@@ -0,0 +1,281 @@
+    """
+    # Make a raster model grid
+    mg = RasterModelGrid((3, 3))
+    _ = mg.add_zeros("soil__depth", at="node")
+    _ = mg.add_zeros("soil_production__rate", at="node")
+    _ = mg.add_zeros("topographic__elevation", at="node")
```
There's no need to include the underscores (here and elsewhere).
```python
    mg.add_zeros("soil__depth", at="node")
    mg.add_zeros("soil_production__rate", at="node")
    mg.add_zeros("topographic__elevation", at="node")
```

mcflugen commented on tests/components/concentration_tracker/test_concentration_tracker_for_diffusion.py:
```diff
@@ -0,0 +1,281 @@
+        _ = ConcentrationTrackerForDiffusion(mg)
+
+
+def test_input_fields_soil():
+    """
+    ConcentrationTrackerForDiffusion should throw an error when input fields
+    are not provided (soil__depth, soil_production__rate, topographic__elevation)
+    """
+    mg = RasterModelGrid((3, 3))
+    _ = mg.add_zeros("soil__flux", at="link")
+
+    fields_to_add = ["soil__depth", "soil_production__rate", "topographic__elevation"]
+
+    # Instantiate the component
+    for field in fields_to_add:
+        with pytest.raises(FieldError):
+            _ = ConcentrationTrackerForDiffusion(mg)
+        _ = mg.add_zeros(field, at="node")
```
This would be a good place to use a parameterized test.
```python
@pytest.mark.parametrize(
    "required_field", ["soil__depth", "soil_production__rate", "topographic__elevation"]
)
def test_input_fields_soil(required_field):
    """
    ConcentrationTrackerForDiffusion should throw an error when input fields
    are not provided (soil__depth, soil_production__rate, topographic__elevation)
    """
    mg = RasterModelGrid((3, 3))
    mg.add_zeros("soil__flux", at="link")

    mg.add_zeros("soil__depth", at="node")
    mg.add_zeros("soil_production__rate", at="node")
    mg.add_zeros("topographic__elevation", at="node")

    mg.at_node.pop(required_field)
    with pytest.raises(FieldError):
        ConcentrationTrackerForDiffusion(mg)
```

mcflugen commented on tests/components/concentration_tracker/test_concentration_tracker_for_diffusion.py:
```diff
@@ -0,0 +1,281 @@
+
+    _ = ConcentrationTrackerForDiffusion(mg)
+
+    node_fields = [
+        "bedrock_property__concentration",
+        "sediment_property__concentration",
+        "sediment_property_decay__rate",
+        "sediment_property_production__rate",
+    ]
+
+    assert "sediment_property__mass_flux" in mg.at_link
+    for node_field in node_fields:
+        assert node_field in mg.at_node
```
```python
missing_fields = {
    "bedrock_property__concentration",
    "sediment_property__concentration",
    "sediment_property_decay__rate",
    "sediment_property_production__rate",
} - set(mg.at_node)
assert not missing_fields
```
might be a little cleaner and will report all the missing fields at once.

mcflugen commented on tests/components/concentration_tracker/test_concentration_tracker_for_diffusion.py:
```diff
@@ -0,0 +1,281 @@
+        _ = ConcentrationTrackerForDiffusion(mg, concentration_in_bedrock=-1)
+
+
+# %% Test against analytical solutions
+# PLACEHOLDER: Test results against 1-D analytical solution (for DepthDependentDiffuser)
+# (I think this is covered by the docstring tests, so I haven't added it here)
+
+# PLACEHOLDER: Test results against 1-D analytical solution (for DepthDependentTaylorDiffuser)
+# (I think this is covered by the docstring tests, so I haven't added it here)
```
These are great ideas! I would like to see more tests that show the component is working correctly. Another test to include would be one that ensures mass balance.

loroberge commented:
@mcflugen I have addressed all of your comments and added some tests to check against simple analytical solutions. 

I have also changed the component calculation in one simple way: the weathering portion of the calculation now uses C_w (concentration_from_weathering) instead of C_br (concentration_in_bedrock) to reflect the possibility that some concentration is gained or lost in the weathering process (i.e., not all soils have exactly the same properties as their parent bedrock). I add this as an optional input to the component which defaults to 'None' (i.e., 'off'). If 'None', the Setter points C_w to C_br and the calculation is done exactly as before. If defined by the user, then the Setter will use the user's value instead of pointing to C_br.

Let me know if there's anything else required from me.

mcflugen commented on landlab/components/concentration_tracker/concentration_tracker_for_diffusion.py:
```diff
@@ -0,0 +1,390 @@
+        self._QC_links = np.zeros(self._grid.number_of_links)
+
+    @property
+    def C_init(self):
+        """Initial concentration in soil/sediment (kg/m^3)."""
+        return self._C_init
+
+    @property
+    def C_br(self):
+        """Concentration in bedrock (kg/m^3)."""
+        return self._C_br
+
+    @property
+    def C_w(self):
+        """Concentration from the weathering process (kg/m^3)."""
```
What do you think about changing the names of these properties to something more descriptive?

Maybe `conc_init`, `conc_bedrock`, `conc_weathering`? Or maybe `concentration_` instead of `conc_` but maybe that's too long?

loroberge replied:
I've gone for conc_ for brevity.

mcflugen commented on landlab/components/concentration_tracker/concentration_tracker_for_diffusion.py:
```diff
@@ -0,0 +1,390 @@
+            raise ValueError("Concentration cannot be negative")
+        self._C_w = new_val
+
+    def concentration(self, dt):
```
How about rename this to be something like `calc_concentration`? Using the `calc_` prefix is a convention we've used elsewhere to indication a calculation is being done rather than just returning a value.

mcflugen commented on landlab/components/concentration_tracker/concentration_tracker_for_diffusion.py:
```diff
@@ -0,0 +1,390 @@
+        dt_over_depth = np.divide(dt, self._soil__depth, where=is_soil)
+        dt_over_depth[~is_soil] = 0.0
+
+        C_local = C_old * old_depth_over_new
+        C_from_weathering = np.divide(
```
Are you ok with changing these to `c_local` and `c_old`? Variable names should be *snake_case*. Notice that GitHub highlights them as though they are classes because of the capital *C*.

loroberge replied:
I have changed all "C" to "conc" across the board for consistency.

mcflugen replied:
Perfect. Thank you.

mcflugen reviewed (commented):
@loroberge Just a few more comments.

mcflugen commented:
@loroberge Thank you! I made just a few more minor comments.

I made a couple of changes to your *docstring* that are mostly just formatting. One thing you might want to look at though is where I've changed things from,
```pycon
>>> np.allclose(
...     mg.at_node["topographic__elevation"].reshape(mg.shape),
...     [
...         [0.0, 4.0, 8.0, 12.0, 16.0],
...         [0.0, 4.11701964, 8.01583689, 11.00247875, 16.0],
...         [0.0, 4.0, 8.0, 12.0, 16.0],
...     ],
... )
True
```
to
```pycon
>>> mg.at_node["topographic__elevation"].reshape(mg.shape)
array([[ 0. ,   4.        ,   8.        ,  12.        , 16. ],
       [ 0. ,   4.11701964,   8.01583689,  11.00247875, 16. ],
       [ 0. ,   4.        ,   8.        ,  12.        , 16. ]])
```
I think it's easier for a user to read this way. If we discover the tests begin to fail due to roundoff error, we could skip those tests by adding ``# doctest: +SKIP`` to those lines. I assume you've just copied the actual output from those calculations to be the expected output so they're really just smoke tests, anyway. What do you think?

loroberge commented:
> @loroberge Thank you! I made just a few more minor comments.
> If we discover the tests begin to fail due to roundoff error, we could skip those tests by adding `# doctest: +SKIP` to those lines. I assume you've just copied the actual output from those calculations to be the expected output so they're really just smoke tests, anyway. What do you think?

Sounds good. I agree that the _docstring_ is much easier to read now. 

One last thing for your review: I've added two tests for mass balance in the test file. One with all boundaries closed and the other with one open boundary across which concentration is fluxed.

mcflugen commented:
> One last thing for your review: I've added two tests for mass balance in the test file. One with all boundaries closed and the other with one open boundary across which concentration is fluxed.

@loroberge Those mass balance checks look great! Thank you for adding them.
